ReactQt/runtime/src/communication/websocketexecutor.cpp

```diff
@@ -0,0 +1,197 @@
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#ifdef RCT_DEV
+
+#include "websocketexecutor.h"
+
+#include <QDebug>
+#include <QJsonArray>
+#include <QJsonDocument>
+#include <QJsonObject>
+#include <QNetworkAccessManager>
+#include <QNetworkReply>
+#include <QNetworkRequest>
+#include <QQueue>
+#include <QTimer>
+#include <QWebSocket>
+
+namespace {
+const int CHROME_DEV_TOOL_AWAIT_TIMEOUT = 5000;
+}
+
+class WebSocketExecutorPrivate : public QObject {
+public:
+    WebSocketExecutorPrivate(WebSocketExecutor* e) : q_ptr(e) {}
+
+    bool processRequests();
+    void processRequest(const QByteArray& request,
+                        const IExecutor::ExecuteCallback& callback = IExecutor::ExecuteCallback());
+    void setupWebSocket(const QUrl& url);
+    QByteArray getMessageData(const QVariantMap& messageData);
+
+public slots:
+    void readReply(const QString& reply);
+    void passReceivedDataToCallback(const QByteArray& data);
+
+public:
+    QQueue<QByteArray> m_requestQueue;
+    QQueue<IExecutor::ExecuteCallback> m_responseQueue;
+    QByteArray m_inputBuffer;
+    WebSocketExecutor* q_ptr = nullptr;
+    QWebSocket m_webSocket;
+    QVariantMap m_injectedObjects;
+    bool m_readyToSendAppScriptData = false;
+    QByteArray m_applicationScriptData;
+};
+
+WebSocketExecutor::WebSocketExecutor(const QUrl& url, QObject* parent)
+    : IExecutor(parent), d_ptr(new WebSocketExecutorPrivate(this)) {
+    Q_D(WebSocketExecutor);
+
+    d->setupWebSocket(url);
+}
+
+WebSocketExecutor::~WebSocketExecutor() {}
+
+void WebSocketExecutorPrivate::setupWebSocket(const QUrl& url) {
+
+    connect(&m_webSocket, &QWebSocket::connected, [=]() {
+        QUrl startDevToolsURL = url.resolved(QStringLiteral("launch-js-devtools"));
+        startDevToolsURL.setScheme("http");
+        QNetworkAccessManager* networkAccessManager = new QNetworkAccessManager();
+        QNetworkReply* reply = networkAccessManager->get(QNetworkRequest(startDevToolsURL));
+
+        connect(reply, &QNetworkReply::readyRead, [=]() {
+            if (reply->readAll() == QStringLiteral("OK")) {
+                QTimer* timer = new QTimer;
+                timer->setSingleShot(true);
+                timer->setInterval(CHROME_DEV_TOOL_AWAIT_TIMEOUT);
+
+                QObject::connect(timer, &QTimer::timeout, [=]() {
+                    QByteArray data = getMessageData(QVariantMap{{"method", "prepareJSRuntime"}});
+                    processRequest(data, [=](const QJsonDocument&) {
+                        m_readyToSendAppScriptData = true;
+                        q_ptr->executeApplicationScriptOnSocketReady();
+                    });
+                    timer->deleteLater();
+                });
+                timer->start();
+            }
+            reply->deleteLater();
```

Comment:
From Qt documetation about QNetworkReply::readyRead: `This signal is emitted once every time new data is available for reading from the device's current read channel`.
That could result in double deletion of `reply` and `networkAccessManager`